File: report/src/main/java/com/hartwig/actin/report/datamodel/ReportFactory.java
package com.hartwig.actin.report.datamodel;

import com.hartwig.actin.algo.datamodel.TreatmentMatch;
import com.hartwig.actin.clinical.datamodel.ClinicalRecord;
import com.hartwig.actin.molecular.datamodel.MolecularRecord;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;

public final class ReportFactory {

    private static final Logger LOGGER = LogManager.getLogger(ReportFactory.class);

    private ReportFactory() {
    }

    @NotNull
    public static Report fromInputs(@NotNull ClinicalRecord clinical, @NotNull MolecularRecord molecular,
            @NotNull TreatmentMatch treatmentMatch) {
        if (!clinical.patientId().equals(treatmentMatch.patientId())) {
            LOGGER.warn("Clinical patientId '{}' not the same as treatment match patientId '{}'! Using clinical patientId",
                    clinical.patientId(),
                    treatmentMatch.patientId());
        }

        return ImmutableReport.builder()
                .patientId(clinical.patientId())
                .sampleId(molecular.sampleId())
                .clinical(clinical)
                .molecular(molecular)
                .treatmentMatch(treatmentMatch)
                .build();
    }
}

File: report/src/main/kotlin/com/hartwig/actin/report/datamodel/ReportFactory.kt
package com.hartwig.actin.report.datamodel

import com.hartwig.actin.datamodel.PatientRecord
import com.hartwig.actin.datamodel.algo.TreatmentMatch
import io.github.oshai.kotlinlogging.KotlinLogging
import java.time.LocalDate

object ReportFactory {

    private val logger = KotlinLogging.logger {}

    fun create(reportDate: LocalDate, patientRecord: PatientRecord, treatmentMatch: TreatmentMatch): Report {
        if (patientRecord.patientId != treatmentMatch.patientId) {
            logger.warn { "Patient record patientId '${patientRecord.patientId}' not the same as treatment match patientId '${treatmentMatch.patientId}'! Using patient record patientId" }
        }

        return Report(
            reportDate = reportDate,
            patientId = patientRecord.patientId,
            patientRecord = patientRecord,
            treatmentMatch = treatmentMatch
        )
    }
}